src/libraries/System.Private.CoreLib/src/System/Math.cs

```diff
@@ -1476,5 +1478,182 @@ public static double ScaleB(double x, int n)
+                valMid = BigMul(val2High, (uint)left);
+            }
+            else
+            {
+                if (val2High != 0)
+                    goto Overflow;
+                // Compute the 'middle' bits of the long multiplication
+                valMid = BigMul(val1High, (uint)right);
+            }
+
+            // See if any bits after bit 32 are set
+            if (High32Bits(valMid) != 0)
+                goto Overflow;
+
+            long ret = (long)(BigMul((uint)left, (uint)right) + (valMid << 32));
+
+            // check for overflow
+            if (High32Bits((ulong)ret) < (uint)valMid)
+                goto Overflow;
+
+            if (sign >= 0)
+            {
+                // have we spilled into the sign bit?
+                if (ret < 0)
+                    goto Overflow;
+            }
+            else
+            {
+                ret = -ret;
+                // have we spilled into the sign bit?
+                if (ret > 0)
+                    goto Overflow;
+            }
+
+#if DEBUG
+            Debug.Assert(ret == result, $"Multiply overflow got: {ret}, expected: {result}");
+#endif
+            return ret;
+
+        Overflow:
+            ThrowHelper.ThrowOverflowException();
+            return 0;
+        }
+
+        [StackTraceHidden]
+        internal static ulong MultiplyUInt64Checked(ulong left, ulong right)
+        {
+            // Get the upper 32 bits of the numbers
+            uint val1High = High32Bits(left);
+            uint val2High = High32Bits(right);
+
+            ulong valMid;
+
+            if (val1High == 0)
+            {
+                if (val2High == 0)
+                    return (ulong)(uint)left * (uint)right;
+                // Compute the 'middle' bits of the long multiplication
+                valMid = BigMul(val2High, (uint)left);
+            }
+            else
+            {
+                if (val2High != 0)
+                    goto Overflow;
+                // Compute the 'middle' bits of the long multiplication
+                valMid = BigMul(val1High, (uint)right);
+            }
+
+            // See if any bits after bit 32 are set
+            if (High32Bits(valMid) != 0)
+                goto Overflow;
+
+            ulong ret = BigMul((uint)left, (uint)right) + (valMid << 32);
+
+            // check for overflow
+            if (High32Bits(ret) < (uint)valMid)
+                goto Overflow;
+
+            Debug.Assert(ret == left * right, $"Multiply overflow got: {ret}, expected: {left * right}");
+            return ret;
+
+        Overflow:
+            ThrowHelper.ThrowOverflowException();
+            return 0;
+        }
+
+        private const double Int32MaxValueOffset = (double)int.MaxValue + 1;
+        private const double UInt32MaxValueOffset = (double)uint.MaxValue + 1;
+
+        [StackTraceHidden]
+        internal static int ConvertToInt32Checked(double value)
```

Comment:
Does the JIT support overloading these? That is, would it work as expected if we eventually defined `internal static int ConvertToInt32Checked(float value)` as well, or do they need to be more uniquely named (like `ConvertDoubleToInt32Checked`?)

Reply:
Yes, we can overload these. It takes an extra ceremony - the lookups have to specify signature to disambiguate.

I was a bit lazy with MultiplyChecked and encoded the type in the name even though the overloading would be more natural pattern. I will rename MultiplyInt64/UInt64Checked to MultiplyChecked, and do the lookup using signature instead.